### ISSUE TITLE
Windows platform fails to draw correctly many Unicode characters

### ISSUE THREAD

**Issue by ManoloFLTK** [fltk/fltk#478]
Program test/utf8 shows that several Unicode characters are not correctly drawn under the Windows platform. See for example the section U+2230 - U+23FF :

![default](https://user-images.githubusercontent.com/41016272/183937224-6e325a29-cb35-421f-b3cc-6ce412755c84.jpg)

I find that changing the _character set_ argument of the `CreateFont()` call in file 
`src/drivers/GDI/Fl_GDI_GraphicsDriver.cxx` from `DEFAULT_CHARSET `to `GB2312_CHARSET` produces a much wider Unicode support :

![GB2312](https://user-images.githubusercontent.com/41016272/183933294-d94435c7-01cf-4033-9bc9-b31c400cb62b.jpg)

Any comment about whether this change is adequate or suggestion for a better code change ?

**Comment by Albrecht-S** [comment]
I don't know what would be adequate or even better ... or "correct" by any means. FWIW, the code in question is in [src/drivers/GDI/Fl_GDI_Graphics_Driver_font.cxx](https://github.com/fltk/fltk/blob/b8c227a8f2bb004de5819c93be5b7f6218402716/src/drivers/GDI/Fl_GDI_Graphics_Driver_font.cxx#L223) (line 223). Picking an arbitrary value `GB2312_CHARSET` doesn't seem right (IMHO).

Slightly unrelated: we should probably convert the font name to wide characters and use [CreateFontW()](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-createfontw) rather than CreateFont() just in case the font _name_ is not bare ASCII.

The following thoughts are only from reading the docs which are pretty hard to understand or interpret (IMHO) unless you're already familiar with the topics (which I am NOT).

The documentation of `CreateFontW()` (see link above) states:
> `DEFAULT_CHARSET` is set to a value based on the current system locale. [...] **To ensure consistent results when creating a font, do not specify OEM_CHARSET or DEFAULT_CHARSET.** If you specify a typeface name in the `lpszFace` parameter, make sure that the `fdwCharSet` value matches the character set of the typeface specified in `lpszFace`.

Note: they don't say what the `fdwCharSet` value is, you should probably know this already? (see below)

Another way to create a font would be [CreateFontIndirectW()](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-createfontindirectw) which documents:

> To get the appropriate font on different language versions of the OS, call [EnumFontFamiliesEx](https://docs.microsoft.com/en-us/windows/desktop/api/wingdi/nf-wingdi-enumfontfamiliesexw) with the desired font characteristics in the [LOGFONT](https://docs.microsoft.com/en-us/windows/desktop/api/wingdi/ns-wingdi-logfontw) structure, retrieve the appropriate typeface name, and create the font using [CreateFont](https://docs.microsoft.com/en-us/windows/desktop/api/wingdi/nf-wingdi-createfontw) or [CreateFontIndirect](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-createfontindirectw).

`EnumFontFamiliesExW()` can obviously be used to find all fonts of a **particular font name**:

> **lfFaceName** : [...] If set to a valid typeface name, the function  enumerates all fonts with the specified name.

_Summary / conclusion:_

Does this help? I don't know. According to the docs we should NOT use `DEFAULT_CHARSET` but using `GB2312_CHARSET` doesn't seem right either (GB2312 is a simplified Chinese character set). Using `EnumFontFamiliesExW()` with the given font name (converted to Windows "wide" characters) could help to find the correct charset value.

However, this is far beyond my "Windows comfort zone" and I'm not able to pursue this myself. Maybe opening a topic in fltk.coredev could reach more devs and may be helpful...

**Comment by Albrecht-S** [comment]
FTR: I noticed that @ManoloFLTK changed `CreateFont()` to  `CreateFontW()` in commit d325fa504e4a4e2a456a646a4159a43a8dfac4fd as suggested. Thank you, Manolo.

**Comment by Albrecht-S** [comment]
Just to add another observation: the new "Unicode Boxes" test in `test/unittests` doesn't draw correctly on Windows as well. This can be seen in `wine` on Linux as well on native Windows (in my VirtualBox instance), no matter if built with MinGW or VS.
Screenshot from `wine`:
![image](https://user-images.githubusercontent.com/26235500/210635992-dd7baa15-cef0-4a8a-854f-0cdf0b9a2bb3.png)

There seems to be a character width issue (Windows/Visual Studio looks similar). I'm not sure if this is a font issue on my system and I didn't test if the suggested patch would help. Just reporting this for later review.

**Comment by ManoloFLTK** [comment]
> I'm not sure if this is a font issue on my system and I didn't test if the suggested patch would help. 

I tried and saw it doesn't help.

**Comment by MatthiasWM** [comment]
So I did a little experiment. Und Windows 11 with GDI code, I get 
![Screenshot 2023-01-05 at 16 36 22](https://user-images.githubusercontent.com/653993/210820358-96ae6468-5392-4988-ba30-b5d590e8f741.jpg)
I thought that using GDIplus may give better results, so here is GDIplus `DrawString`:
![Screenshot 2023-01-05 at 16 35 01](https://user-images.githubusercontent.com/653993/210820609-15c8aabc-862b-42b1-a21c-2723482e9033.jpg)
For giggles, I set the `StringFormatFlagsNoFontFallback` flag to see if there are some character fallbacks happening, ad yes, there are (empty rectangles in the screenshot below):
![Screenshot 2023-01-05 at 16 35 35](https://user-images.githubusercontent.com/653993/210821190-637b4061-8018-4b3e-a4fc-5eea7b0194c7.jpg)
So what we are seeing here is some missing characters for the thin outline box. I tried this with Lucida Console as the fixed font, and things are only aligned when the fallback is *disabled*. With the fallback enabled, we get the thin boxes, but the spaces are still too small.

So in conclusion, when fallback to another font is enabled, the thin outline characters are replaced with those from a font that is wider than the selected font. I know that Windows allows us to give a list of fallback fonts for every font we choose, so we could prioritise those, but man, this gets very complicated. GDIplus alone is not the solution.

**Comment by rageworx** [comment]
Dear @ManoloFLTK , @MatthiasWM, and @Albrecht-S 
I am really interested in this issue too ...
I have tested with some Windows fonts, and found it seems to depend on what supported code in TTF/OTF.

![image](https://github.com/fltk/fltk/assets/6948225/74a98160-ab31-45bf-918a-9af19fca2d59)

I can see most of rendered words for some default true type fonts, but some others are not.

![image](https://github.com/fltk/fltk/assets/6948225/d705eb47-6811-48d3-9538-cba951620b7d)

Maybe possible to change configuration for DEFAULT_CHARSET by like `fltk_system_charset()` ?

**Comment by ManoloFLTK** [comment]
@Albrecht-S @MatthiasWM I now propose a less intrusive change to improve on this topic: change the font used behind FL_HELVETICA under Windows and use "Microsoft Sans Serif" instead of "Arial".
The new font is very very close in shape and sizes to the previous one as shown here:

<img width="1105" alt="Image" src="https://github.com/user-attachments/assets/4b52604d-9d62-4afa-b52a-c86ca2232591" />
which ensures application GUI layouts won't be hardly impacted.

The new font covers many more Unicode points:

<img width="580" alt="Image" src="https://github.com/user-attachments/assets/1828ae5a-1f75-4ed0-93dc-93920bdb404a" />

Since FL_HELVETICA is FLTK's default font for most purposes, the new font would be widely used by FLTK apps and would  cover Unicode better than previously.

@rageworx  The question of how to ask Windows to use an alternate font when a Unicode point not covered by the active font appears remains unanswered, unfortunately.

Here is the proposed patch to change the FL_HELVETICA font under Windows:
```
diff --git a/src/drivers/GDI/Fl_GDI_Graphics_Driver_font.cxx b/src/drivers/GDI/Fl_GDI_Graphics_Driver_font.cxx
index c5a6e0308..fccb90917 100644
--- a/src/drivers/GDI/Fl_GDI_Graphics_Driver_font.cxx
+++ b/src/drivers/GDI/Fl_GDI_Graphics_Driver_font.cxx
@@ -280,10 +280,10 @@ Fl_GDI_Font_Descriptor::~Fl_GDI_Font_Descriptor() {
 // WARNING: if you add to this table, you must redefine FL_FREE_FONT
 // in Enumerations.H & recompile!!
 static Fl_Fontdesc built_in_table[] = {
-{" Arial"},
-{"BArial"},
-{"IArial"},
-{"PArial"},
+  {" Microsoft Sans Serif"},//{" Arial"},
+  {"BMicrosoft Sans Serif"},//{"BArial"},
+  {"IMicrosoft Sans Serif"},//{"IArial"},
+  {"PMicrosoft Sans Serif"},//{"PArial"},
 {" Courier New"},
 {"BCourier New"},
 {"ICourier New"},
```

**Comment by Albrecht-S** [comment]
Looks good. Two questions came to my mind:

1. Is this font available on all our supported Windows versions? The answer is "yes" although in very different versions, as [documented here](https://learn.microsoft.com/en-us/typography/font-list/microsoft-sans-serif). That's good.

2. Could there be unwanted effects, for instance because users are comparing font **names**? I'm not sure if this is an issue though.

[Wikipedia](https://en.wikipedia.org/wiki/Microsoft_Sans_Serif#Description) elaborates on the number of glyphs that are supported (and which ones) by the different font versions and that this font is also available on macOS (since 10.5).

That said, I'm `+1` 👍 on changing the font but I wonder if this should be done now or better in 1.5 development. Comments welcome.

**Comment by ManoloFLTK** [comment]
> Could there be unwanted effects, for instance because users are comparing font names? 

I don't see an issue here either because to access font via their names, FLTK requires a distinct API than to access them via their symbolic names FL_HELVETICA, FL_TIMES, … So, I would believe apps use one or the other API in different contexts and won't compare a font name to a symbolic name. The list of available font names is unaffected by the proposed change.

**Comment by ManoloFLTK** [comment]
This change of font has been committed (915ea80).

@Albrecht-S I leave you find where to further document that, if needed.

**Comment by Albrecht-S** [comment]
> @Albrecht-S I leave you find where to further document that, if needed.

OK, will do ...
Thanks for your work on this and the commit.
